notes/lecture14.tex

```diff
@@ -0,0 +1,273 @@
+\begin{align}
+x ^T = ((x^{(1)})^T, \cdots, (x^{(N)})^T) &\qquad& x_i \in \R^{n_i}, \sum_{i=1}^N n_i = n \\
+A = [A_1 \vert \cdots \vert A_N] &\qquad& A x = \sum_{i=1}^N A_i x^{(i)} \\
+f(x) = \sum_{i=1}^N f_i(x^{(i)})
+\end{align}
+
+Then the Lagrangian is also separable in $x$
+\[
+L(x, \lambda) = \sum_{i=1}^N \left( f_i(x^{(i)}) + \lambda^T A_i x^{(i)} - \frac{1}{N} \lambda^T b\right) = \sum_{i=1}^N L_i(x^{(i)}, \lambda)
+\]
+
+Each term in the sum consists of one non-interacting partition
+$(x^{(i)}, A_i, f_i)$, so minimization of each term in the sum can
+occur in parallel. This leads to the \emph{dual decomposition
+algorithm}:
+
+\begin{itemize}
+\item
+  In parallel on worker nodes:
+  $x_t^{(i)} = \arg\inf_{x^{(i)}} L_i(x^{(i)}, \lambda_t)$
+\item
+  On a master node: $\lambda_{t+1} = \lambda_t + \eta (A x - b)$
+\end{itemize}
+
+\begin{example}[Consensus optimization]
+Consensus optimization is an
+application that comes up in distributed computing which can be solved
+using dual decomposition. Given a graph $G = (V, E)$,
+\[
+\min_x \sum_{v \in V} f_v(x_v) : x_v = x_u~\forall(u,v) \in E
+\]
+This problem is separable over $v \in V$, so dual decomposition applies.
+
+\end{example}
+
+\begin{example}[Network utility maximization]
+Suppose we have a
+network with $k$ links, each with capacity $c_i$. We are interested
+in allocating $N$ different flows with fixed routes over these links such that
+utility is maximized and resource constraints are not exceeded.
+Let $x_i \in \R^N$ represent the amount of flow $i$ allocated and
+$U_i : \R \to \R$ a convex utility function which
+returns the amount of utility obtained from having $x_i$ amount of
+flow $i$. The optimization problem is
+\[
+\max_x \sum_{i=1}^N U_i(x_i) : R x \leq c
+\]
+where $R$ is a $k \times N$ matrix whose $(k,i)$th entry gives the
+amount of the capacity of link $k$ is consumed by flow $i$.
+
+To rewrite the primal problem in standard form (i.e.~as a minimization),
+take negatives:
+\[
+\min_x -\sum_i U_i(x^{(i)}) : R x \leq c
+\]
+
+The dual problem is then
+\[
+\max_{\lambda \geq 0} \min_x \sum_i -U_i(x^{(i)}) + \lambda^T (R x - c)
+\]
+where the $R x \leq c$ primal inequality constraint results in the $\lambda \geq 0$ constraint.
+The second term can be rewritten as $\lambda^T\left(\sum_i R_i x_i - \frac{1}{N}c\right)$,
+showing that the dual splits over $i$ and hence dual decomposition applies. This leads to a
+parallel algorithm which each worker node computes
+\[
+\arg\max_{x_i} U_i(x_i) - \lambda^T R_i x_i
+\]
+and the master node computes
+\[
+\lambda_{t+1} = \left[\lambda_t + \eta (R x - c)\right]_+
+\]
+We take the positive part because of the $\lambda \geq 0$ constraint.
+
+\textbf{Aside}: In resource allocation problems, the values of the dual variables
+$\lambda$ at the optimal point have an economic interpretation as
+``prices'' to the resources. In this example, $\lambda_k$ can be
+interpreted as the utility per unit of flow over link $k$.
```

Comment:
should be "interpreted as the price per unit of flow'

Reply:
:+1: